metabase/metabase
# Changing visualization settings for custom columns triggers browser native warning message

Closes https://github.com/metabase/metabase/issues/30905

### Description

The bug was noticed [here](https://stats.metabase.com/question/9703).

Currently, when visualization settings for custom columns are changed, the Query Builder store updates the `card` and `originalCard` to different column setting values, which comes from a discrepancy between `getColumnKey` and the `ExpressionDimension`, where `getColumnKey` would return an expression field_ref of `["expression", "EXPRESSION_NAME"]` while `ExpressionDimension` would return a value of `["expression", "EXPRESSION_NAME", options]` where `options` might be a object or `null` (but in any case the array would have a length of 3). More concretely, this means that:

`state > qb > card` would be updated with:
```
"column_settings": {
    "["ref",["expression","EXPRESSION_NAME", null]]": {
     "show_mini_bar": false
   }
  }
```

while `state > qb > originalCard` would be updated with:
```
  "column_settings": {
    "["ref",["expression","EXPRESSION_NAME"]]": {
     "show_mini_bar": false
  }
```

The `isDirty` flag in the Query Builder uses a hash diff between the `card` and `originalCard`, so even though the expressions produce the same result, they're not equal and `isDirty` would always be `true`. So, when a user leaves a question that they didn't changed or had just saved, the browser still warns the user when they try to leave the question, and the `Save` button is also still there.

After speaking with @calherries, I learned that expressions should not have options. So, I went into `getColumnKey` and `ExpressionDimension` to make sure that all expression `field_ref` values always had a length of 2 and didn't have any options.

**Looking at when we are using expressions in our codebase, I notice that there are places where expressions _do_ have options. I'll need some guidance on whether or not we should keep those, or if there are places where expression options are necessary.**

_Edit: Ok, from what has been mentioned in various places throughout Slack, expressions shouldn't have `null` within the options parameter. I've removed references to `null` within the `Dimension` and `getColumnKey` tests, since the tests will fail if they expect `["expression", "expression_name", null]` when they should expect `["expression", "expression_name"]`_


### How to verify

To reproduce the bug:

1. Create a new question with a custom column
2. Make any changes to the visualization settings for that custom column - add a mini bar, change the formatting, whatever
3. Save the question

Then you’ll see:
- The Save button is still there
- Refreshing the page triggers the warning

With this PR, after creating a custom column, changing the vis settings, and saving, the `Save` button should disappear and refreshing the page should not produce a browser warning.

### Demo

_Upload a demo video or before/after screenshots if sensible or remove the section_

### Checklist
